### ISSUE TITLE
Exclude GPL hgal dependency

### ISSUE THREAD

**Issue by joshsh** [robstewart57/rdf4h#93]
rdf4h is currently used in a data transformation toolkit called [Dragon](https://eng.uber.com/dragon-schema-integration-at-uber-scale/). We are preparing to release Dragon as open-source software, but have run into a snag with respect to the rdf4h dependency, which includes a transitive dependency on hgal. While rdf4h is BSD-licensed, hgal uses GPL v2, which unfortunately prevents its inclusion in Dragon.

I see that hgal is only needed for one isolated function in Data.RDF.Query (isGraphIsomorphic). If this function is not seen as essential, would it be possible to remove the hgal dependency in a future release of rdf4h?

**Comment by robstewart57** [comment]
HI @joshsh I've asked the hgal maintainer, who'd like to keep the GPL v2 license for that library.

One option therefore could be for me to create a gist, copy/pasting the `isGraphIsomorphic` function there with the GitHub page pointing to it if people want to compute RDF graph isomorphism using hgal. Then, remove the `isGraphIsomorphic` from the rdf4h library and the testsuite, along with the hgal dependency.

**Comment by joshsh** [comment]
Thanks for checking, @robstewart57 . Yes, if you are willing to downgrade that function to a Gist and remove the dependency, that would allow me (and potentially others with license constraints) to proceed.

**Comment by robstewart57** [comment]
@joshsh OK, I've moved the `isGraphIsomorphic` function out to a gist, and I've removed `hgal` from the cabal file.

This is reflected in release 4.0.2.

Removing `isGraphIsomorphic` in the unit tests, instead relying on `isIsomorphic`, identified a small handful of Turtle and XML parser bugs in rdf4h, and I will fix these when I return in a couple of months.

**Comment by joshsh** [comment]
I see the compile errors. Separate issue but do you know if there is a way I can ignore the errors when depending on rdf4h via Stack? Otherwise, I will still have to come up with an rdf4h alternative for our initial release.

**Comment by joshsh** [comment]
For now, I have copied the sources I need into Dragon, with appropriate license information. When the build issues are resolved, I will remove these sources and restore the rdf4h dependency. Thanks.

**Comment by robstewart57** [comment]
@joshsh Compiler errors? There aren't any compiler errors in version 4.0.2 or git HEAD, as far as I'm aware?

What I referred to with:

>  identified a small handful of Turtle and XML parser bugs in rdf4h

Is some unit tests fail with `stack test`. `stack build` and `stack install` should compile fine.

Let me know if this isn't the case for you.

**Comment by joshsh** [comment]
Now that you mention it, yes rdf4h 4.0.2 builds from source without any issues. The errors I encountered were due to dependency conflicts, and I was able to resolve them. I will issue a PR to remove hgal from rdf4h's stack.yaml, as FOSSA may notice it and flag it as a GPL dependency otherwise.

**Comment by robstewart57** [comment]
Thanks, I'll accept the PR once you've submitted it.

**Comment by joshsh** [comment]
Actually, FOSSA does not flag this. I guess it only has access to the .cabal file. There is no longer anything standing in the way of a release of Dragon with rdf4h. Thanks!

**Comment by robstewart57** [comment]
@joshsh I've removed it from `stack.yaml` anyway, since it is no longer needed.

https://github.com/robstewart57/rdf4h/commit/46cdc7e49068831cce890a6584e03b7eb5bad144
